File: solutions/JavaScript/345.ReverseVowelsOfAString.js
var reverseVowels = function (s) {
  const vowels = new Set(['a', 'e', 'i', 'o', 'u']);
  const sArray = s.split('');

  let left = 0;
  let right = s.length - 1;

  while (left < right) {
    if (!vowels.has(sArray[left].toLowerCase())) {
      left++;
    } else if (!vowels.has(sArray[right].toLowerCase())) {
      right--;
    } else {
      [sArray[left], sArray[right]] = [sArray[right], sArray[left]];
      left++;
      right--;
    }
  }

  return sArray.join('');
};

File: solutions/Python/345.ReverseVowelsOfAString.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = {"a", "e", "i", "o", "u"}
        s_list = list(s)

        left = 0
        right = len(s) - 1

        while left < right:
            if s_list[left].lower() not in vowels:
                left += 1
            elif s_list[right].lower() not in vowels:
                right -= 1
            else:
                s_list[left], s_list[right] = s_list[right], s_list[left]
                left += 1
                right -= 1

        return "".join(s_list)
